### ISSUE TITLE
ILLEGAL_TYPE_OF_ARGUMENT allow_nullable_key + final

### ISSUE THREAD

**Issue by den-crane** [ClickHouse/ClickHouse#54156]
https://fiddle.clickhouse.com/796d6997-b571-4e4f-aede-1e18f670ec6f

```sql
CREATE TABLE t
(
    d Nullable(Date),
    f1 Nullable(String),
    f2 Nullable(String),
    c Nullable(Int64)
)
ENGINE = ReplacingMergeTree()
ORDER BY (f1, f2, d)
SETTINGS allow_nullable_key = 1;

insert into t select today() d, 
  [number%999999, null][number%2] f1,  
  ['x', null][number%2] f2, 
  [number, null][number%2] c 
from numbers(1000000);

SELECT date_trunc('month', d), SUM( c )  
FROM t FINAL WHERE f2 = 'x' GROUP BY 1;

Received exception from server (version 23.7.5):
Code: 43. DB::Exception: Received from localhost:9000. 
DB::Exception: Illegal types of arguments (Date, UInt16) of function greater: 
While processing NOT ((f1, f2, d) > ('203217', 'x', 19600)). (ILLEGAL_TYPE_OF_ARGUMENT)
```

cc @amosbird

**Comment by tavplubix** [comment]
cc: @amosbird

**Comment by amosbird** [comment]
The test case unveils another issue (incorrect query result), which is also fixed in #54164
